### PR TITLE
DAOS-10141 vos: init vo_max_write for upgrading case

### PR DESCRIPTION
**Pull request by wangshilong**
If object creating before following change:
89809(DAOS-9053 vos: Track accurate latest write time for an object (#7845))

Which could happen while upgrading DAOS from 2.0 to 2.2.

It is possible @vos_max_write is zero with an existed object,
assign current epoch to avoid later assertion.

Signed-off-by: Wang Shilong <shilong.wang@intel.com>

### CODE REVIEW

**Review by jolivier23 — DISMISSED** [review]
Where is the assertion?

**Comment by jolivier23** [comment]
I believe max_write is in 2.0.  Do we need to backport this to keep 2.0.0 compatible with 2.0.2?

**Comment by wangshilong** [comment]
> 



> Where is the assertion?

src/vos/vos_aggregate.c:259 need_aggregate() Assertion 'entry->ie_last_update != 0' failed

**Comment by wangshilong** [comment]
> I believe max_write is in 2.0. Do we need to backport this to keep 2.0.0 compatible with 2.0.2?

I think max_write was not in 2.0? i hit this problem when upgrading DAOS from 2.0 to 2.2

**Review by NiuYawei — DISMISSED** [review]

**Comment by jolivier23** [comment]
> > I believe max_write is in 2.0. Do we need to backport this to keep 2.0.0 compatible with 2.0.2?
> 
> I think max_write was not in 2.0? i hit this problem when upgrading DAOS from 2.0 to 2.2

You are right.  I thought it had landed but it has not.

File: src/vos/vos_obj_index.c [hunk 1]
```diff
@@ -233,8 +233,12 @@ vos_oi_find_alloc(struct vos_container *cont, daos_unit_oid_t oid,
 		DP_UOID(oid));
 
 	rc = vos_oi_find(cont, oid, &obj, ts_set);
-	if (rc == 0)
+	if (rc == 0) {
+		/* upgrading case, set it to current epoch */
+		if (obj->vo_max_write == 0)
```
Comment (jolivier23) [inline src/vos/vos_obj_index.c:238]: Does this always work?   This code only executes if we hold the object which we don't do during iteration

Comment (wangshilong) [reply src/vos/vos_obj_index.c]: You are right, after some more testing, i think codes should be fixed in oi_iter_fetch() directly where ie_last_update is inited.

File: src/vos/vos_obj_index.c [hunk 2]
```diff
@@ -623,7 +623,11 @@ oi_iter_fetch(struct vos_iterator *iter, vos_iter_entry_t *it_entry,
 	}
 	it_entry->ie_child_type = VOS_ITER_DKEY;
 
-	it_entry->ie_last_update = obj->vo_max_write;
+	/* Upgrading case, set it to current epoch */
+	if (obj->vo_max_write == 0)
+		it_entry->ie_last_update = epr.epr_hi;
```
Comment (NiuYawei) [inline src/vos/vos_obj_index.c:628]: This is incorrect, you'd call vos_ilog_last_update() to get the last update epoch.

**Review by NiuYawei — CHANGES_REQUESTED** [review]

**Review by NiuYawei — APPROVED** [review]

**Review by jolivier23 — APPROVED** [review]

**Pull request merged**